src/report/s3.rs

```diff
@@ -0,0 +1,100 @@
+use errors::*;
+use mime::Mime;
+use report::ReportWriter;
+use rusoto_core::{DefaultCredentialsProvider, Region, default_tls_client};
+use rusoto_s3::{PutObjectRequest, S3, S3Client};
+use std::borrow::Cow;
+use std::fmt::{self, Display};
+use std::io;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use uri::Uri;
+
+#[derive(Debug, Clone)]
+pub struct S3Prefix {
+    pub bucket: String,
+    pub prefix: PathBuf,
+}
+
+impl FromStr for S3Prefix {
+    type Err = Error;
+    fn from_str(uri: &str) -> Result<S3Prefix> {
+        match Uri::new(uri).chain_err(|| ErrorKind::BadS3Uri)? {
```

Comment:
I'm not sure if it is better to pull in `uri` in addtion to `url` for this, but it seemed easier to assert that there are no extraneous parts to a `Uri` as compared to a `Url`.